File: services/platform-api/src/handlers/pilot.rs
use axum::Json;
use axum::extract::State;
use axum::http::{HeaderMap, header};
use chrono::{Duration, Utc};
use serde::Deserialize;
use serde_json::{Value, json};
use sha2::{Digest, Sha256};
use sqlx::Row;

use crate::AppState;
use crate::types::*;

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct BootstrapRequest {
    pub token: String,
}

pub async fn bootstrap(
    State(state): State<AppState>,
    headers: HeaderMap,
    Json(req): Json<BootstrapRequest>,
) -> Result<(HeaderMap, Json<Value>), ApiError> {
    // 1. Origin verification for pilot session bootstrap (exact allowlist when
    // CORS_ALLOWED_ORIGINS is configured; presence-only in dev)
    crate::auth::require_allowed_origin(&headers, &state.jwt_config.pilot_allowed_origins)?;

    // 2. Hash the bootstrap invitation token
    let mut invite_hasher = Sha256::new();
    invite_hasher.update(req.token.as_bytes());
    let invite_hash = format!("{:x}", invite_hasher.finalize());

    // 3. Atomically verify and consume the invitation
    let mut tx = state
        .pool
        .begin()
        .await
        .map_err(|e| ApiError::Database(e.to_string()))?;

    let row =
        sqlx::query("SELECT tenant_id, learner_id FROM app.consume_pilot_invitation_by_hash($1)")
            .bind(&invite_hash)
            .fetch_optional(&mut *tx)
            .await?;

    let Some(row) = row else {
        return Err(ApiError::Unauthorized);
    };

    let tenant_id: String = row.try_get("tenant_id")?;
    let learner_id: String = row.try_get("learner_id")?;

    // Set local tenant_id for RLS and query user profile (bound set_config, matching
    // begin_tenant_tx — never string-interpolate into SQL)
    sqlx::query("SELECT set_config('app.tenant_id', $1, true)")
        .bind(&tenant_id)
        .execute(&mut *tx)
        .await?;

    let user_row =
        sqlx::query("SELECT display_name, role FROM users WHERE id = $1 AND tenant_id = $2")
            .bind(&learner_id)
            .bind(&tenant_id)
            .fetch_one(&mut *tx)
            .await?;

    let display_name: String = user_row.try_get("display_name")?;
    let role: String = user_row.try_get("role")?;

    if role != "learner" {
        return Err(ApiError::Forbidden);
    }

    // 4. Generate new session token, session ID, and CSRF token
    let session_token = uuid::Uuid::new_v4().to_string();
    let session_id = next_id("pilot-session");
    let csrf_token = uuid::Uuid::new_v4().to_string();

    let mut session_hasher = Sha256::new();
    session_hasher.update(session_token.as_bytes());
    let session_hash = format!("{:x}", session_hasher.finalize());

    let now = Utc::now();
    let idle_expires_at = now + Duration::hours(8);
    let absolute_expires_at = now + Duration::hours(24);

    sqlx::query(
        "INSERT INTO pilot_sessions (id, tenant_id, learner_id, token_hash, csrf_token, idle_expires_at, absolute_expires_at)
         VALUES ($1, $2, $3, $4, $5, $6, $7)"
    )
    .bind(&session_id)
    .bind(&tenant_id)
    .bind(&learner_id)
    .bind(&session_hash)
    .bind(&csrf_token)
    .bind(idle_expires_at)
    .bind(absolute_expires_at)
    .execute(&mut *tx)
    .await?;

    // Create bootstrap audit event
    let audit_id = next_id("audit");
    sqlx::query(
        "INSERT INTO audit_events (id, tenant_id, actor_id, action, subject_type, subject_id)
         VALUES ($1, $2, $3, 'pilot.session.bootstrap', 'user', $3)",
    )
    .bind(&audit_id)
    .bind(&tenant_id)
    .bind(&learner_id)
    .execute(&mut *tx)
    .await?;

    tx.commit().await?;

    // Set cookie headers
    let mut resp_headers = HeaderMap::new();
    let cookie_val = format!(
        "__Host-qrai-pilot={}; HttpOnly; Secure; SameSite=Strict; Path=/; Max-Age=28800",
        session_token
    );
    resp_headers.insert(
        header::SET_COOKIE,
        cookie_val
            .parse()
            .map_err(|e: axum::http::header::InvalidHeaderValue| {
                ApiError::BadRequest(format!("Failed to parse cookie: {e}"))
            })?,
    );

    Ok((
        resp_headers,
        Json(json!({
            "userId": learner_id,
            "tenantId": tenant_id,
            "displayName": display_name,
            "role": role,
            "csrfToken": csrf_token,
        })),
    ))
}

pub async fn logout(
    State(state): State<AppState>,
    headers: HeaderMap,
) -> Result<(HeaderMap, Json<Value>), ApiError> {
    // Look up the cookie to find the active session to revoke
    if let Some(cookie_header) = headers.get("cookie")
        && let Ok(cookie_str) = cookie_header.to_str()
    {
        for cookie in cookie_str.split(';') {
            let cookie = cookie.trim();
            if let Some(val) = cookie.strip_prefix("__Host-qrai-pilot=") {
                let mut hasher = Sha256::new();
                hasher.update(val.as_bytes());
                let session_hash = format!("{:x}", hasher.finalize());

                // Find session using SECURITY DEFINER function to bypass initial RLS
                let row = sqlx::query(
                    "SELECT id, tenant_id, learner_id FROM app.get_pilot_session_by_hash($1)",
                )
                .bind(&session_hash)
                .fetch_optional(&state.pool)
                .await?;

                if let Some(row) = row {
                    let session_id: String = row.try_get("id")?;
                    let tenant_id: String = row.try_get("tenant_id")?;
                    let learner_id: String = row.try_get("learner_id")?;

                    // Revoke session under a tenant transaction
                    let mut tx = crate::begin_tenant_tx(&state.pool, &tenant_id).await?;
                    sqlx::query("UPDATE pilot_sessions SET revoked_at = now() WHERE id = $1")
                        .bind(&session_id)
                        .execute(&mut *tx)
                        .await?;

                    // Audit logout
                    let audit_id = next_id("audit");
                    sqlx::query(
                        "INSERT INTO audit_events (id, tenant_id, actor_id, action, subject_type, subject_id)
                         VALUES ($1, $2, $3, 'pilot.session.logout', 'user', $3)"
                    )
                    .bind(&audit_id)
                    .bind(&tenant_id)
                    .bind(&learner_id)
                    .execute(&mut *tx)
                    .await?;

                    tx.commit().await?;
                }
            }
        }
    }

    // Set cookie clearing headers
    let mut resp_headers = HeaderMap::new();
    let cookie_val = "__Host-qrai-pilot=; HttpOnly; Secure; SameSite=Strict; Path=/; Max-Age=0; Expires=Thu, 01 Jan 1970 00:00:00 GMT";
    resp_headers.insert(
        header::SET_COOKIE,
        cookie_val
            .parse()
            .map_err(|e: axum::http::header::InvalidHeaderValue| {
                ApiError::BadRequest(format!("Failed to parse cookie: {e}"))
            })?,
    );

    Ok((resp_headers, Json(json!({ "status": "logged_out" }))))
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MintInvitationRequest {
    pub learner_id: String,
    /// Validity window in hours; defaults to 168 (7 days), clamped to [1, 720] so an admin
    /// cannot mint an effectively immortal invite.
    pub ttl_hours: Option<i64>,
}

/// Admin/Ops mint a SINGLE-USE pilot invitation for a learner IN THEIR OWN TENANT (P1.6).
///
/// The raw token is returned exactly once — only its SHA-256 hash is stored, so a leaked DB row
/// cannot be replayed into a session. The learner opens `?invite=<token>`; the web app exchanges
/// it for a `__Host-qrai-pilot` cookie via `bootstrap`. This is the only way pilot invitations are
/// issued, closing the "invitation-issuance mechanism not present anywhere" gap from research.md.
pub async fn mint_invitation(
    State(state): State<AppState>,
    method: axum::http::Method,
    headers: HeaderMap,
    Json(req): Json<MintInvitationRequest>,
) -> Result<Json<Value>, ApiError> {
    let actor = crate::auth::resolve_actor(&method, &headers, &state).await?;
    actor.require_any(&[ActorRole::Admin, ActorRole::Ops])?;

    let ttl_hours = req.ttl_hours.unwrap_or(168).clamp(1, 720);

    let mut tx = crate::begin_tenant_tx(&state.pool, &actor.tenant_id).await?;

    // The invited user must exist in the caller's tenant (RLS scopes the lookup) and be a learner —
    // the pilot cookie pins role=learner, so inviting a teacher/admin would silently under-privilege
    // them and is almost certainly a mistake; reject it explicitly rather than mint a dead invite.
    let target = sqlx::query("SELECT role FROM users WHERE id = $1 AND tenant_id = $2")
        .bind(&req.learner_id)
        .bind(&actor.tenant_id)
        .fetch_optional(&mut *tx)
        .await?;
    let Some(target) = target else {
        return Err(ApiError::NotFound);
    };
    let target_role: String = target.try_get("role")?;
    if target_role != "learner" {
        return Err(ApiError::BadRequest(
            "pilot invitations may only target learner accounts".to_string(),
        ));
    }

    // Raw token shown once; store only its hash (mirrors bootstrap's session-token handling).
    let token = uuid::Uuid::new_v4().to_string();
    let mut hasher = Sha256::new();
    hasher.update(token.as_bytes());
    let token_hash = format!("{:x}", hasher.finalize());

    let invitation_id = next_id("pilot-invitation");
    let expires_at = Utc::now() + Duration::hours(ttl_hours);

    sqlx::query(
        "INSERT INTO pilot_invitations (id, tenant_id, learner_id, token_hash, expires_at)
         VALUES ($1, $2, $3, $4, $5)",
    )
    .bind(&invitation_id)
    .bind(&actor.tenant_id)
    .bind(&req.learner_id)
    .bind(&token_hash)
    .bind(expires_at)
    .execute(&mut *tx)
    .await?;

    let audit_id = next_id("audit");
    sqlx::query(
        "INSERT INTO audit_events (id, tenant_id, actor_id, action, subject_type, subject_id)
         VALUES ($1, $2, $3, 'pilot.invitation.minted', 'user', $4)",
    )
    .bind(&audit_id)
    .bind(&actor.tenant_id)
    .bind(&actor.user_id)
    .bind(&req.learner_id)
    .execute(&mut *tx)
    .await?;

    tx.commit().await?;

    // ponytail: per-request env read on a rare admin endpoint (not an auth decision, just a URL
    // prefix); move to AppState if invitation minting ever becomes hot. Absent base => caller builds
    // the URL from the raw token itself, which is the source of truth.
    let invite_url = std::env::var("PILOT_INVITE_BASE_URL")
        .ok()
        .filter(|b| !b.is_empty())
        .map(|base| format!("{}/?invite={}", base.trim_end_matches('/'), token));

    Ok(Json(json!({
        "invitationId": invitation_id,
        "learnerId": req.learner_id,
        "token": token,
        "inviteUrl": invite_url,
        "expiresAt": expires_at.to_rfc3339(),
    })))
}

File: server/src/routes/pilot.mjs
/**
 * N13b — the three pilot-session routes. Port of handlers/pilot.rs.
 *
 * The cookie ATTRIBUTES are the contract here, not just the value. `__Host-` is a browser-enforced
 * prefix: a cookie with that name is rejected outright unless it is `Secure`, `Path=/`, and carries
 * no `Domain`. Dropping any one of them does not produce an error anywhere — the browser silently
 * refuses to store it and the learner simply never stays logged in.
 */
import { createHash, randomUUID } from "node:crypto";

import { ApiError, Forbidden, NotFound, RejectionError, Unauthorized, requireAnyRole, requireAllowedOrigin, resolveActor, pilotCookieToken } from "../lib/authz.mjs";
import { fractional } from "./progress.mjs";
import { proxy } from "../lib/proxy.mjs";

const sha256Hex = (s) => createHash("sha256").update(s, "utf8").digest("hex");

/** pilot.rs:118 — every attribute is load-bearing; see the module comment. */
const SESSION_COOKIE = (token) =>
  `__Host-qrai-pilot=${token}; HttpOnly; Secure; SameSite=Strict; Path=/; Max-Age=28800`;

/** pilot.rs:197 — the clearing cookie carries BOTH Max-Age=0 and an Expires in the past. */
const CLEAR_COOKIE =
  "__Host-qrai-pilot=; HttpOnly; Secure; SameSite=Strict; Path=/; Max-Age=0; Expires=Thu, 01 Jan 1970 00:00:00 GMT";

/** POST /v1/pilot/session/bootstrap — pilot.rs:19 */
export async function bootstrap(req, reply, ctx) {
  // Origin FIRST, before the token is even hashed. This endpoint is reachable with no credentials
  // at all — the invitation token IS the credential — so the Origin check is the only thing
  // standing between it and a cross-site page redeeming a token it phished.
  requireAllowedOrigin(req.headers, ctx.pilotAllowedOrigins);

  const token = req.body?.token;
  if (typeof token !== "string") throw new RejectionError("token is required", 422);

  const body = await ctx.db.withDiscoveredTenant(async (tx) => {
    // The tenant is not known until this locked-down SECURITY DEFINER function consumes the token.
    const [row] = await tx`
      SELECT tenant_id, learner_id FROM app.consume_pilot_invitation_by_hash(${sha256Hex(token)})`;
    // 401, not 404: missing, expired, and already-consumed invitations are the same answer.
    if (!row) throw Unauthorized("invitation not valid");
    return { tenantId: row.tenant_id, learnerId: row.learner_id };
  }, async (tx, { tenantId, learnerId }) => {
    // `withDiscoveredTenant` installed the transaction-local GUC and shared statement timeout
    // before this tenant-owned work became callable.
    const [user] = await tx`
      SELECT display_name, role FROM users WHERE id = ${learnerId} AND tenant_id = ${tenantId}`;
    if (!user) throw new ApiError("invited user not found", 500);

    // The pilot cookie pins role=learner. Minting a session for anyone else would silently
    // under-privilege them, so it is refused rather than quietly downgraded.
    if (user.role !== "learner") throw Forbidden("pilot sessions are for learners only");

    const sessionToken = randomUUID();
    const sessionId = `pilot-session-${randomUUID()}`;
    const csrfToken = randomUUID();
    const now = Date.now();

    await tx`
      INSERT INTO pilot_sessions
        (id, tenant_id, learner_id, token_hash, csrf_token, idle_expires_at, absolute_expires_at)
      VALUES (${sessionId}, ${tenantId}, ${learnerId}, ${sha256Hex(sessionToken)}, ${csrfToken},
              ${new Date(now + 8 * 3600 * 1000)}, ${new Date(now + 24 * 3600 * 1000)})`;

    // The learner is BOTH actor and subject here: nobody else acted. `$3` is bound twice in the
    // Rust for the same reason.
    await tx`
      INSERT INTO audit_events (id, tenant_id, actor_id, action, subject_type, subject_id)
      VALUES (${`audit-${randomUUID()}`}, ${tenantId}, ${learnerId},
              'pilot.session.bootstrap', 'user', ${learnerId})`;

    return {
      sessionToken,
      // json! keys, alphabetical.
      payload: {
        csrfToken,
        displayName: user.display_name,
        role: user.role,
        tenantId,
        userId: learnerId,
      },
    };
  });

  reply.header("set-cookie", SESSION_COOKIE(body.sessionToken));
  return reply.send(body.payload);
}

/** POST /v1/pilot/session/logout — pilot.rs:142 */
export async function logout(req, reply, ctx) {
  const token = pilotCookieToken(req.headers.cookie);

  if (token !== null && token !== "") {
    const [row] = await ctx.db.sql`
      SELECT id, tenant_id, learner_id FROM app.get_pilot_session_by_hash(${sha256Hex(token)})`;

    if (row) {
      await ctx.db.withTenant(row.tenant_id, async (tx) => {
        await tx`UPDATE pilot_sessions SET revoked_at = now() WHERE id = ${row.id}`;
        await tx`
          INSERT INTO audit_events (id, tenant_id, actor_id, action, subject_type, subject_id)
          VALUES (${`audit-${randomUUID()}`}, ${row.tenant_id}, ${row.learner_id},
                  'pilot.session.logout', 'user', ${row.learner_id})`;
      });
    }
  }

  // ALWAYS 200 with the clearing cookie, even when no session was found. Logout must be idempotent
  // and must never reveal whether the cookie it was handed was real — a 401 here tells a caller
  // that a stolen cookie has already been revoked.
  reply.header("set-cookie", CLEAR_COOKIE);
  return reply.send({ status: "logged_out" });
}

/** POST /v1/pilot/invitations — pilot.rs:225 */
export async function mintInvitation(req, reply, ctx) {
  const resolved = await resolveActor(req, ctx);
  if (resolved.delegate) return proxy(req, reply, ctx.upstream);
  const { actor } = resolved;

  requireAnyRole(actor, ["admin", "ops"]);

  const learnerId = req.body?.learnerId;
  if (typeof learnerId !== "string") throw new RejectionError("learnerId is required", 422);

  // Default 7 days, clamped to [1, 720] so an admin cannot mint an effectively immortal invite.
  const requested = req.body?.ttlHours;
  const ttlHours = Math.min(Math.max(typeof requested === "number" ? Math.trunc(requested) : 168, 1), 720);

  const body = await ctx.db.withTenant(actor.tenantId, async (tx) => {
    const [target] = await tx`
      SELECT role FROM users WHERE id = ${learnerId} AND tenant_id = ${actor.tenantId}`;
    if (!target) throw NotFound();
    // A teacher or admin would be silently under-privileged by a cookie that pins role=learner, so
    // this is a 400 naming the rule rather than a dead invitation nobody can use.
    if (target.role !== "learner") {
      throw new ApiError("pilot invitations may only target learner accounts", 400);
    }

    const token = randomUUID();
    const invitationId = `pilot-invitation-${randomUUID()}`;
    const expiresAt = new Date(Date.now() + ttlHours * 3600 * 1000);

    // Only the HASH is stored. The raw token is shown exactly once, so a leaked DB row cannot be
    // replayed into a session.
    await tx`
      INSERT INTO pilot_invitations (id, tenant_id, learner_id, token_hash, expires_at)
      VALUES (${invitationId}, ${actor.tenantId}, ${learnerId}, ${sha256Hex(token)}, ${expiresAt})`;

    await tx`
      INSERT INTO audit_events (id, tenant_id, actor_id, action, subject_type, subject_id)
      VALUES (${`audit-${randomUUID()}`}, ${actor.tenantId}, ${actor.userId},
              'pilot.invitation.minted', 'user', ${learnerId})`;

    // chrono's to_rfc3339() again — `+00:00`, not `Z`, with 0/3/6 fractional digits. Formatted in
    // Postgres for the same reason as every other timestamp in this port.
    const [{ base, us }] = await tx`
      SELECT to_char(${expiresAt}::timestamptz AT TIME ZONE 'UTC', 'YYYY-MM-DD"T"HH24:MI:SS') AS base,
             (EXTRACT(microseconds FROM ${expiresAt}::timestamptz)::bigint % 1000000) AS us`;

    const base_url = process.env.PILOT_INVITE_BASE_URL;
    return {
      expiresAt: `${base}${fractional(Number(us))}+00:00`,
      invitationId,
      // "invitationId" < "inviteUrl": at index 5 it is 'a' vs 'e'. Alphabetical, not eyeballed.
      inviteUrl: base_url ? `${base_url.replace(/\/+$/, "")}/?invite=${token}` : null,
      learnerId,
      token,
    };
  });

  return reply.send(body);
}
